File: Hard/2026-08-17-1563-Stone-Game-V/solution.java
class Solution {
    public int stoneGameV(int[] stoneValue) {
        int n = stoneValue.length;
        int[] prefixSum = new int[n + 1];
        for (int i = 0; i < n; i++) {
            prefixSum[i + 1] = prefixSum[i] + stoneValue[i];
        }
        
        int[][] dp = new int[n][n];
        
        for (int length = 2; length <= n; length++) {
            for (int i = 0; i <= n - length; i++) {
                int j = i + length - 1;
                for (int k = i; k < j; k++) {
                    int leftSum = prefixSum[k + 1] - prefixSum[i];
                    int rightSum = prefixSum[j + 1] - prefixSum[k + 1];
                    if (leftSum < rightSum) {
                        dp[i][j] = Math.max(dp[i][j], dp[i][k] + leftSum);
                    } else if (leftSum > rightSum) {
                        dp[i][j] = Math.max(dp[i][j], dp[k + 1][j] + rightSum);
                    } else {
                        dp[i][j] = Math.max(dp[i][j], Math.max(dp[i][k] + leftSum, dp[k + 1][j] + rightSum));
                    }
                }
            }
        }
        
        return dp[0][n - 1];
    }
}

File: Hard/2026-08-17-1563-Stone-Game-V/solution.py
class Solution:
    def stoneGameV(self, stoneValue: List[int]) -> int:
        n = len(stoneValue)
        prefix_sum = [0] * (n + 1)
        for i in range(n):
            prefix_sum[i + 1] = prefix_sum[i] + stoneValue[i]
        
        dp = [[0] * n for _ in range(n)]
        
        for length in range(2, n + 1):
            for left in range(n - length + 1):
                right = left + length - 1
                for mid in range(left, right):
                    left_sum = prefix_sum[mid + 1] - prefix_sum[left]
                    right_sum = prefix_sum[right + 1] - prefix_sum[mid + 1]
                    if left_sum < right_sum:
                        dp[left][right] = max(dp[left][right], dp[left][mid] + left_sum)
                    elif left_sum > right_sum:
                        dp[left][right] = max(dp[left][right], dp[mid + 1][right] + right_sum)
                    else:
                        dp[left][right] = max(dp[left][right], dp[left][mid] + left_sum, dp[mid + 1][right] + right_sum)
        
        return dp[0][n - 1]